### PR TITLE
TDD for gluu-passport

### PR DESCRIPTION
**Pull request by ldeveloperl1985**
- Mock gluu endpoints and get/set configuration
- Added test case for openidconnect-default type
- To test other flow, need to add provider data in `config-data.js` 
- global `expect` and `assert`
- updated `package.json`, separate `prod` and `dev` depedencies.

### CODE REVIEW

File: test/config.test.js [hunk 1]
```diff
@@ -1,8 +1,72 @@
...
+fs.writeFileSync(passportConfigResponse.conf.spTLSCert, passportSPCert);
+fs.writeFileSync(passportConfigResponse.conf.spTLSKey, passportSPKey);
+
+/**
+ * root level hooks
+ */
+before((done) => {
+	mockIDP();
+	// init and get/set all config
+	global.app = require('../server/app');
```
Comment (christian-hawk) [inline test/config.test.js:40]: Nice @ldeveloperl1985 !
Learning a lot here. 🧑‍🏫

File: test/config.test.js [hunk 2]
```diff
@@ -25,5 +89,65 @@ describe('productioncfg', function() {
...
+	const reqheaders = {
+		authorization: `Bearer ${rptTokenResponse.access_token}`,
+		pct: rptTokenResponse.pct
+	};
+	nock(idpURL, {
+		reqheaders
+	})
+		.get(idpPassportConfigPath)
+		.reply(200, passportConfigResponse);
+}
```
Comment (christian-hawk) [inline test/config.test.js:153]: Just an additional comment, this config.test.js was initially created to test the configuration files (`default.js`, `production.js`, `test.js`). Maybe I didn't let it clear. Gonna add this comment to the file in my next commit.

File: server/routes.js [hunk 3]
```diff
@@ -11,6 +11,10 @@ const
 	logger = require('./utils/logging'),
 	path = require('path')
 
+router.get('/health-check', function(req, res) {
```
Comment (christian-hawk) [inline server/routes.js:14]: cool, I would never thought of it!

File: test/openid-connect-default-type.test.js [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+describe('Openidconnect-default type passport flow', () => {
+	it('Authorization request to external IDP', (done) => {
+		request(app)
+			.get('/passport/token')
+			.expect(200)
+			.end((err, tokenResponse) => {
+                if (err) return done(err);
+                
+                expect(200).to.equal(tokenResponse.statusCode);
```
Comment (christian-hawk) [inline test/openid-connect-default-type.test.js:11]: I love this one!!! 😄 👍

**Comment by christian-hawk** [comment]
I merged the last branch to resolve conflict from the #103

File: config/test.js [hunk 5]
```diff
@@ -1,26 +1,152 @@
-const passportFile = {
```
[the remark this answers is not in the document]
Comment (ldeveloperl1985) [reply config/test.js:1]: Agreed, I like this idea. I'll update it

File: test/config.test.js [hunk 6]
```diff
@@ -25,5 +89,65 @@ describe('productioncfg', function() {
 		assert.exists(
 			productioncfg.saltFile, 'saltFile is not null or undefined')
 	})
-})
+});
 
+function mockIDP() {
```
[the remark this answers is not in the document]
Comment (ldeveloperl1985) [reply test/config.test.js:94]: Agree, I'll refactor code

File: test/openid-connect-default-type.test.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
+const request = require('supertest');
+
+describe('Openidconnect-default type passport flow', () => {
+	it('Authorization request to external IDP', (done) => {
```
[the remark this answers is not in the document]
Comment (ldeveloperl1985) [reply test/openid-connect-default-type.test.js:4]: Agree, it should be OP, I'll update it

File: server/app.js [hunk 8]
```diff
@@ -13,7 +13,7 @@ const
 	confDiscovery = require('./utils/configDiscovery'),
 	routes = require('./routes'),
 	providers = require('./providers'),
-	passportFile = config.get('passportFile')
+	passportFile = process.env.passport_config_file || config.get('passportFile')
```
[the remark this answers is not in the document]
Comment (ldeveloperl1985) [reply server/app.js:16]: Got it, I'll update it

File: server/app.js [hunk 9]
```diff
@@ -94,7 +95,7 @@ function pollConfiguration(configEndpoint) {
 			logger.log2('debug', e.stack)
 			logger.log2('warn', 'An attempt to get configuration data will be tried again soon')
 		})
-	setTimeout(pollConfiguration, 60000, configEndpoint)	 //1 minute timer
+	setTimeout(pollConfiguration, process.env.config_update_timer || 60000, configEndpoint)	 //1 minute timer
```
[the remark this answers is not in the document]
Comment (ldeveloperl1985) [reply server/app.js:98]: Agree

File: config/test.js [hunk 10]
```diff
@@ -1,26 +1,152 @@
...
-	failureRedirectUrl:
-	'https://chris.gluuthree.org/oxauth/auth/passport/passportlogin.htm',
-
-	logLevel: 'debug',
-
-	consoleLogOnly: false,
+const saltFile = './test/testdata/salt'
 
-	clientId: '1502.d49baf9f-b19b-40de-a990-33d08e7f9e77',
+// data of passportFile
```
[the remark this answers is not in the document]
Comment (ldeveloperl1985) [reply config/test.js:3]: I'll refactor the code and move all other config in helper.js

File: test/config.test.js (hunk 2 above)
Comment (ldeveloperl1985) [reply test/config.test.js]: Please check new changes, I moved all other config in `helper.js`

**Pull request merged**